docs/cs-proof.md

```diff
@@ -0,0 +1,800 @@
+
+We will use the notation described in the [`notation`](../notes/index.html#notation) section of the notes. In addition, we rename two more variables from the paper for better consistency with the rest of the notation:
+\\[
+\begin{aligned}
+    \beta         &\xrightarrow{} \tilde{o} \\\\
+	Q             &\xrightarrow{} q \\\\
+\end{aligned}
+\\]
+
+[bp_website]: https://crypto.stanford.edu/bulletproofs/
+
+Constraint system
+-----------------
+
+The constraint system consists of two sets of constraints: [multiplication gates](#multiplication-gates) and [linear constraints](#linear-constraints) in terms of the [variables](#variables).
+
+### Variables
+
+There are two kinds of variables in the constraint system:
+
+1. \\(m\\) **high-level** variables, the input secrets \\(\mathbf{v}\\),
+2. \\(n\\) **low-level** variables, the internal inputs and outputs of the [multiplication gates](#multiplication-gates).
+
+### Multiplication gates
+
+Each multiplication gate takes two input variables and multiplies them to get an output.
+That relation for all multiplication gates is represented by \\(n\\) constraints:
+
+\\[
+\mathbf{a}\_L \circ \mathbf{a}\_R = \mathbf{a}\_O,
+\\]
+where:
+* \\(\mathbf{a}\_L\\) is the vector of the first input to each gate,
+* \\(\mathbf{a}\_R\\) is the vector of the second input to each gate,
+* \\(\mathbf{a}\_O\\) is the vector of multiplication results.
+
+### Linear constraints
+
+Linear constraints are expressed using a vector of \\(q\\) equations that use linear combinations of the [variables](#variables):
+
+\\[
+\mathbf{W}\_L \cdot \mathbf{a}\_L +
+\mathbf{W}\_R \cdot \mathbf{a}\_R +
+\mathbf{W}\_O \cdot \mathbf{a}\_O =
+\mathbf{W}\_V \cdot \mathbf{v} +
+\mathbf{c},
+\\]
+where:
+
+* \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O\\) are weights applied to the respective inputs and outputs of the [multiplication gates](#multiplication-gates) (low-level [variables](#variables)),
+* \\(\mathbf{W}\_V\\) are weights applied to the high-level [variables](#variables) \\(\mathbf{v}\\),
+* \\(\mathbf{c}\\) is a vector of constant terms used in the linear constraints.
+
+
+### Building constraints
+
+Bulletproofs framework allows building constraint systems _on the fly_, without a trusted setup.
+This allows instantiating constraints from a _family of constraint systems_ parametrized by
+public values and commitments to the secret inputs.
+As a result, the instantiation can be thought of as a _challenge_ generated by a verifier to a prover.
+
+The prover starts out by committing to its secret inputs \\(\mathbf{v}\\)
+and obtaining \\(m\\) _variables_ representing these inputs.
+
+Then, the prover creates linear combinations of available [variables](#variables)
```

Comment:
The prover very rarely adds linear combinations of only high-level variables. (you could, but it doesn't make a lot of sense...) Since we already have "2. Add a constraint" in the later steps, maybe take out this line?

Reply:
👍